feat: event type pbac

sean-brydon opened the pull request:
Seed your database by "yarn seed-pbac"

Look at the roles created in orgs/teams roles and permissions in settings. 

Create new roles/edit old ones to remove all permission from event type - assign it to a user.
Impersonate the user and try to CRUD an event type.
Tick on permissions in the role and keep testing the event type. 


Impersonate a team without pbac enabled and ensure everything works as expected

sean-brydon committed:
fix members page crash with pbac feature flag

sean-brydon committed:
invite member to org backend

sean-brydon committed:
update org permissions

sean-brydon committed:
feat: org profile update settings

sean-brydon committed:
org general page

sean-brydon committed:
remove redudant me call

sean-brydon committed:
privacy page

sean-brydon committed:
add attributes to pbac

sean-brydon committed:
dync + sso on organization permissions

sean-brydon committed:
add tests for resource-permission util

sean-brydon committed:
pass permissions to attributes\

sean-brydon committed:
restore invite members

sean-brydon committed:
update org update and attribute backends

sean-brydon committed:
fix type errors

sean-brydon committed:
fix orgId

sean-brydon committed:
fix types attempt two

sean-brydon committed:
fix types attempt two

sean-brydon committed:
fix merge coflict

sean-brydon committed:
fix type error

sean-brydon committed:
show/hide team event types based on eventType.read permission

sean-brydon committed:
Merge branch 'main' into feat/org-settings-pbac

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
Merge branch 'main' into feat/org-settings-pbac

sean-brydon committed:
Merge branch 'main' into feat/org-settings-pbac

sean-brydon committed:
fix dupe string in i18n

sean-brydon committed:
Merge branch 'main' into feat/org-settings-pbac

sean-brydon committed:
Merge branch 'main' into feat/org-settings-pbac

sean-brydon committed:
fix tests

sean-brydon committed:
fix team-dsync

sean-brydon committed:
update session to use profile

sean-brydon committed:
use profile metadata to get orgRolew

sean-brydon committed:
fix dsync

sean-brydon committed:
fix typing

sean-brydon committed:
Merge branch 'feat/org-settings-pbac' into feat/event-type-pbac

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
Event type create pbac check

sean-brydon committed:
fix the readonly permission check. WIP

sean-brydon committed:
refactor getUserEventTypes

sean-brydon committed:
fix test

sean-brydon committed:
add generate util for update

sean-brydon committed:
update router to use new generator function for pbac procedures

sean-brydon committed:
update transform utils

sean-brydon committed:
fix members being able to update on UI

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
fix allSettled

sean-brydon committed:
fix fallback permissions

sean-brydon committed:
fix permission logic

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
Merge remote-tracking branch 'refs/remotes/origin/feat/event-type-pbac' into feat/event-type-pbac

sean-brydon committed:
fix compare memebrship type error

sean-brydon committed:
fix unit test error

sean-brydon committed:
fix manage test error

eunjae-lee reviewed (commented):
Leaving some comments for now. I need to test it.

sean-brydon commented on packages/trpc/server/routers/viewer/eventTypes/create.handler.ts:
```diff
@@ -89,14 +115,41 @@ export const createHandler = async ({ ctx, input }: CreateOptions) => {
+      throw new TRPCError({ code: "UNAUTHORIZED" });
+    }
+
     const isSystemAdmin = ctx.user.role === "ADMIN";
```
Hmmm - i dont know we if want system admins to have * on every user. Some have impersonation disabled etc. Might need to think about this more to enable it globally

sean-brydon committed:
fix nested permissions

sean-brydon committed:
filter based on canView permission in getUserEventGroups

sean-brydon committed:
Update packages/trpc/server/routers/viewer/eventTypes/utils/EventTypeGroupFilter.test.ts

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
fix tests

sean-brydon committed:
Merge remote-tracking branch 'refs/remotes/origin/feat/event-type-pbac' into feat/event-type-pbac

sean-brydon committed:
fix tests

CarinaWolli commented:
We should probably remove manage already in this PR. It's still showing at the moment: 
<img width="485" height="259" alt="Screenshot 2025-08-12 at 12 30 46 PM" src="https://github.com/user-attachments/assets/ff12edb8-19f7-44e1-a384-0afc917b9ea8" />

CarinaWolli reviewed (commented):
Do we want to include the Workflows and Webhooks tab in this PR as well? That's currently not yet working as it should

Also, another question not related to this PR: Do we plan to also add permissions for Webhooks? I think it would be useful

sean-brydon commented:
> Do we want to include the Workflows
Yes Ill do that 

> Also, another question not related to this PR: Do we plan to also add permissions for Webhooks? I think it would be useful

I hadnt thought about it but yeah we for sure can

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
feat: Adding workflow permission checks inside of eventTypes (#23038)

* add permission checks for workflows

* wip getall active workflows

* check eventType.update permission on activiating workflow on event type

* pass permissions to context and use context in client

keithwillcode committed:
Merge branch 'main' into feat/event-type-pbac

eunjae-lee commented:
I created a new role on a team, but cannot assign it to a team member. I think it used to be possible in this branch. Please correct me if I'm wrong.

btw, I reset the local db, ran seed-pbac, and tested with pbac-owner for role settings.

---

<img width="2814" height="1832" alt="Screenshot 2025-08-28 at 14 49 22@2x" src="https://github.com/user-attachments/assets/802c936f-131e-49da-965d-8abd88c62b81" />


<img width="2820" height="1918" alt="Screenshot 2025-08-28 at 14 49 36@2x" src="https://github.com/user-attachments/assets/65fb8409-8de6-40d8-817e-e7699a1f9d79" />

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
remove local parsePermissionString and use registery

sean-brydon committed:
Use new scoped function

sean-brydon committed:
default scope to orgs

sean-brydon committed:
use object values for workflow permissions

sean-brydon committed:
fix type errors

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon commented on packages/features/pbac/client/context/EventPermissionContext.tsx:
Store for storing permissions across the entire eventType page instead of prop drilling 100 components

sean-brydon commented on packages/trpc/server/routers/viewer/eventTypes/_router.ts:
```diff
@@ -126,32 +127,38 @@ export const eventTypesRouter = router({
 
-  update: eventOwnerProcedure.input(ZUpdateInputSchema).mutation(async ({ ctx, input }) => {
-    const { updateHandler } = await import("./update.handler");
+  update: createEventPbacProcedure("eventType.update", [MembershipRole.ADMIN, MembershipRole.OWNER])
```
Using new factory for PBAC procedure (specific to events here)
 
 Combines old logic plus checking permissions and falling back

sean-brydon commented on packages/trpc/server/routers/viewer/workflows/activateEventType.handler.ts:
```diff
@@ -97,6 +95,21 @@ export const activateEventTypeHandler = async ({ ctx, input }: ActivateEventType
+    const hasPermissionToActivate = await permissionCheckService.checkPermission({
+      userId: ctx.user.id,
+      teamId: eventType.teamId,
+      permission: "eventType.update",
```
This is a workflow handler that activates something on a event type so it makes sense to use eventType.update here

sean-brydon commented on packages/features/pbac/client/hooks/useEventPermission.ts:
```diff
@@ -0,0 +1,51 @@
+import type { EventPermissions } from "../context/EventPermissionContext";
+import { useEventPermissionStore } from "../context/EventPermissionContext";
+
+export const useEventTypePermission = (permission: keyof EventPermissions["eventTypes"]) => {
+  return useEventPermissionStore((state) => ({
+    hasPermission: state.hasEventTypePermission(permission),
+    permissions: state.permissions.eventTypes,
+  }));
+};
```
Zustand does this by default

eunjae-lee reviewed (commented):
code review done. i'm testing at the moment and here is a question:


Q1.

https://github.com/user-attachments/assets/49e9c977-ccb8-4f92-8510-5c8d102edcf8

Q2.

https://github.com/user-attachments/assets/b00f9f85-8a6f-4116-b079-34e6d807dff5

sean-brydon commented:
> code review done. i'm testing at the moment and here is a question:
> 
> Q1.

Interesting issue - will take a look 

> Q2.

This is the drawback i mentiond the other week about removing the "managed" state. To actually remove event-types from a user assigned to a team. You need to ensure they do not have this permission on a higher org scope (which they do by default) so you would need to create a role in orgs with no event types then a role in team with no event types. This being quite a niche usecase for this since private teams exists. After discussion with @CarinaWolli and @ciaranha we thought this is best just left as is without adding complexity

sean-brydon committed:
Fix permission traversal + fix all state caluclation

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

eunjae-lee commented:
Regarding Q2, sorry I didn't share all the information.

I actually created a role on the org level with no event type permissions attached.
Also created a role on the team level with no event type permissions attached.

Then this user should really not see any event types, right? 🤔


https://github.com/user-attachments/assets/e5cb25ca-27cc-4d9d-8401-258d1af61f46

sean-brydon committed:
fix eventType group to use permissions.canRead

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

devin-ai-integration[bot] committed:
fix: update failing PBAC unit tests and resolve ESLint warnings

- Fix getUserEventGroups test: MEMBER role should not have update permissions in fallback scenario
- Fix EventTypeGroupFilter test: add missing canRead property to mock permissions map
- Replace 'as any' type assertions with proper 'as unknown as' type assertions to resolve ESLint warnings

Both tests now pass and align with PBAC permission model where MEMBER role has read-only access to event types.

Co-Authored-By: sean@cal.com <Sean@brydon.io>

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
add pbac to _heavy router

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
Merge main

sean-brydon committed:
restore versions

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

volnei commented on packages/platform/atoms/event-types/wrappers/EventTypeWebWrapper.tsx:
```diff
@@ -29,6 +31,21 @@ import { useEventTypeForm } from "../hooks/useEventTypeForm";
 import { useHandleRouteChange } from "../hooks/useHandleRouteChange";
 import { useTabsNavigations } from "../hooks/useTabsNavigations";
 
+type EventPermissions = {
+  eventTypes: {
+    canRead: boolean;
+    canCreate: boolean;
+    canUpdate: boolean;
+    canDelete: boolean;
+  };
+  workflows: {
+    canRead: boolean;
+    canCreate: boolean;
+    canUpdate: boolean;
+    canDelete: boolean;
+  };
+};
```
I think it makes sense @sean-brydon

volnei commented on packages/features/pbac/client/hooks/useEventPermission.ts:
```diff
@@ -0,0 +1,51 @@
+import { useEventPermissionStore } from "../context/EventPermissionContext";
+
+export const useEventTypePermission = (permission: keyof EventPermissions["eventTypes"]) => {
+  return useEventPermissionStore((state) => ({
```
yeah! Nice!

volnei commented on packages/trpc/server/routers/viewer/eventTypes/heavy/create.handler.ts:
```diff
@@ -50,6 +52,30 @@ export const createHandler = async ({ ctx, input }: CreateOptions) => {
+      hasOrgEventTypeCreatePermission = orgPermissions.canCreate;
+    } catch (error) {
+      // If PBAC check fails, fall back to isOrgAdmin
+      console.warn(`PBAC org check failed for user ${userId}, falling back to isOrgAdmin:`, error);
```
Do we need this console?

volnei reviewed (changes requested):
small comments on code

sean-brydon committed:
Parallelizable members query

sean-brydon committed:
fix nits

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
fix create

sean-brydon committed:
Merge branch 'main' into feat/event-type-pbac

sean-brydon committed:
Restore lock file

sean-brydon committed:
Fix import from merge artifact

volnei commented on packages/trpc/server/routers/viewer/eventTypes/teamAccessUseCase.ts:
```diff
@@ -0,0 +1,47 @@
+          userId,
+          teamId: membership.team.id,
+          permission: "eventType.read",
+          fallbackRoles: ["ADMIN", "OWNER", "MEMBER"], // All roles can read event types by default
```
Can we use Membership constants here?

volnei reviewed (changes requested):
Nit comments, other than that, looks great!

sean-brydon commented:
> Nit comments, other than that, looks great!

Will address these in a follow up! Lets get this merged
